templates/.github/workflows/ci.yml

```diff
@@ -0,0 +1,145 @@
+# .github/workflows/prod-pipeline.yml
+# This is a github action pipeline for production.
+name: ci pipeline
+on:
+  push:
+    branches: [master]
+  pull_request:
+    branches: ['*']
+# Jobs
+jobs:
+  unit-test:
+    runs-on: ubuntu-latest
+    env:
+      CI: true
+      build_env: production
+    steps:
+    # Checkout the code
+    - uses: actions/checkout@v2
+    # Run Tests
+    - name: Run Unit Tests
+      uses: actions/setup-node@v1
+      with:
+        node-version: 14.x
+    - run: |
+        yarn
+        yarn test --watchAll=false
+  build:
+      # Using condition to only builds on `master` branch
+      # Github actions does not have branch option for excluding steps for branches
+      if: github.ref == 'refs/heads/master'
+      needs: unit-test
+      runs-on: ubuntu-latest
+      env:
+        CI: true
+      strategy:
+        matrix:
+          include:
+          - env: stage
+            bucket: <% index .Params `stagingFrontendSubdomain` %><% index .Params `stagingHostRoot` %>
+          - env: production
+            bucket: <% index .Params `productionFrontendSubdomain` %><% index .Params `productionHostRoot` %>
+      outputs:
+        needs-approval: "${{ steps.needs-approval.outputs.needs-approval }}"
+        approved-deploy: "${{ steps.approved-deploy.outputs.approved-deploy }}"
+      steps:
+      # Checkout the code
+        - uses: actions/checkout@v2
+      # Build Code
+        - name: Build Static Site
+          uses: actions/setup-node@v1
+          with:
+            node-version: 12.x
+        - run: |
+            yarn
+            yarn build
+          env:
+            REACT_APP_CONFIG: ${{ matrix.env }}
+      # Upload build to github artifact
+        - name: Upload build artifact to Github
+          uses: actions/upload-artifact@v1
+          with:
+            name: build-artifact-${{ matrix.env }}
+            path: build/
+        ## The outputs set provide data and set matrix for the deploy steps
+        - name: needs-approval
+          id: needs-approval
+          if: ${{ matrix.env == 'production' }}
+          run: echo "::set-output name=needs-approval::{\"include\":[{\"env\":\"${{ matrix.env }}\",\"buildName\":\"build-artifact-${{ matrix.env }}\",\"bucket\":\"${{ matrix.bucket }}\"}]}"
+        - name: approved-deploy
+          id: approved-deploy
+          if: ${{ matrix.env == 'stage' }}
+          run: echo "::set-output name=approved-deploy::{\"include\":[{\"env\":\"${{ matrix.env }}\",\"buildName\":\"build-artifact-${{ matrix.env }}\",\"bucket\":\"${{ matrix.bucket }}\"}]}"
```

Comment:
Using these 2 `outputs` to dictate what gets ran in the following steps (by `strategy:matrix` of the other steps)
So the only declaration of the content is in the build step, and deploy / production-deploy is entirely dynamic

Reply:
Could you add some more documentation, maybe a high level description of how this works? It's quite confusing.